### PR TITLE
Add Container Runtime Interface log parser

### LINKED ISSUE #4616: Add support for Container Runtime Interface (CRI) log format
**Describe the problem:**

Kubernetes uses the CRI log format for container stdout/stderr log streams.

**Additional context**

Per https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/logs/logs.go#L125,  CRI log format example:

```
2016-10-06T00:17:09.669794202Z stdout P log content 1
2016-10-06T00:17:09.669794203Z stderr F log content 2
```

where "log content 1", "log content 2" can be another log format e.g. apache access log.

Archived design proposal: https://github.com/kubernetes/design-proposals-archive/blob/main/node/kubelet-cri-logging.md

### PR DESCRIPTION
**Pull request by sydp**
## One line description of pull request

Adds support for the CRI text log format seen in GKE.

## Description:

**Related issue (if applicable):** fixes #4616

## Notes:
All contributions to Plaso undergo [code review](https://github.com/log2timeline/l2tdocs/blob/main/process/Code%20review%20process.asciidoc).
This makes sure that the code has appropriate test coverage and conforms to the
[Plaso style guide](https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html).

One of the maintainers will examine your code, and may request changes. Check off the items below in
order, and then a maintainer will review your code.

## Checklist:
* [x] Automated checks (GitHub Actions, AppVeyor) pass
* [x] No new [new dependencies](https://plaso.readthedocs.io/en/latest/sources/developer/Adding-a-new-dependency.html) are required or l2tdevtools has been updated
* [x] Reviewer assigned

### CODE REVIEW

**Comment by sydp** [comment]
This is dependent on https://github.com/log2timeline/dfdatetime/pull/264

**Comment by joachimmetz** [comment]
Thanks for the PR, I'll take a look as time permits.

File: plaso/parsers/text_plugins/cri.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    tag (str): the log tag.  Currently only 'P' (partial) and 'F' (full) are
+        supported.
+  """
+  DATA_TYPE = 'cri:container:log:entry'
+
+  def __init__(self):
+    """Initializes event data."""
+    super(CRIEventData, self).__init__(data_type=self.DATA_TYPE)
+    self.event_datetime = None
+    self.message = None
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/cri.py:37]: message is reserved for output field name, please use body or equiv.

File: plaso/parsers/text_plugins/cri.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  ENCODING = 'utf-8'
+
+  _TWO_DIGITS = pyparsing.Word(pyparsing.nums, exact=2)
+
+  _FOUR_DIGITS = pyparsing.Word(pyparsing.nums, exact=4)
+
+  _NINE_DIGITS = pyparsing.Word(pyparsing.nums, exact=9)
+
+  # Date and time values are formatted as: 2016-10-06T00:17:09.669794202Z
+  _DATE_AND_TIME = pyparsing.Combine(
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/cri.py:57]: I recommend to make this a single regex, which appears to be more performant.

File: test_data/cri.log [hunk 3]
```diff
@@ -0,0 +1,2 @@
+2016-10-06T00:17:09.669794202Z stdout P log content 1
```
Comment (joachimmetz) [inline test_data/cri.log:1]: A couple of more example log lines that don't contain sensitive information would be beneficial here

File: tests/parsers/text_plugins/cri.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    expected_event_values = {
+        'data_type': 'cri:container:log:entry',
+        'event_datetime': '2016-10-06T00:17:09.669794203+00:00',
+        'message': ' log content 2',
+        'stream': 'stderr',
+        'tag': 'F'}
+
+    event_data = storage_writer.GetAttributeContainerByIndex('event_data', 1)
+    self.CheckEventData(event_data, expected_event_values)
+
```
Comment (joachimmetz) [inline tests/parsers/text_plugins/cri.py:52]: style nit: add empty line

File: plaso/parsers/text_plugins/cri.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+      key (str): name of the parsed structure.
+      structure (pyparsing.ParseResults): tokens from a parsed log line.
+
+    Raises:
+      ParseError: if the structure cannot be parsed.
+    """
+    if key == 'log_line':
+      golang_time_object = golang_time.GolangTime()
+      golang_time_string = self._GetValueFromStructure(structure, 'date_time')
+      golang_time_object.CopyFromDateTimeString(golang_time_string)
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/cri.py:100]: why convert to golang time ? why not use time elements icw CopyFromStringISO8601 ?

File: test_data/cri.log (hunk 3 above)
Comment (sydp) [reply test_data/cri.log:1]: Done

File: tests/parsers/text_plugins/cri.py (hunk 4 above)
Comment (sydp) [reply tests/parsers/text_plugins/cri.py:52]: Done

File: plaso/parsers/text_plugins/cri.py (hunk 5 above)
Comment (joachimmetz) [reply plaso/parsers/text_plugins/cri.py:100]: Yes please go the TimeElementsInNanoseconds route

Comment (sydp) [reply plaso/parsers/text_plugins/cri.py:100]: I had a closer look and it looks like the time stamps are formatted using "RFC3339Nano" which removes trailing zeros.

https://github.com/containerd/containerd/blob/main/internal/cri/io/logger.go#L38

https://pkg.go.dev/time  

Given this seems specific to golang formatted datetime strings, perhaps adding a `CopyFromRFC3339NanoDateString` to golang_time might be more appropriate?

Comment (joachimmetz) [reply plaso/parsers/text_plugins/cri.py:100]: Why not TimeElementsInNanoseconds with the matching precision? dfDateTime accounts for such difference between storage granularity and value granularity, also see: https://dfdatetime.readthedocs.io/en/latest/sources/Date-and-time-values.html#terminology

This way you don't need add yet another copy function.

Comment (joachimmetz) [reply plaso/parsers/text_plugins/cri.py:100]: Added in https://github.com/log2timeline/dfdatetime/pull/284

File: plaso/parsers/text_plugins/cri.py (hunk 1 above)
Comment (sydp) [reply plaso/parsers/text_plugins/cri.py]: Done

File: plaso/parsers/text_plugins/cri.py (hunk 2 above)
Comment (sydp) [reply plaso/parsers/text_plugins/cri.py]: Done

File: plaso/parsers/text_plugins/cri.py (hunk 5 above)
Comment (sydp) [reply plaso/parsers/text_plugins/cri.py]: Done, thanks for the updates.

**Review by joachimmetz — APPROVED** [review]
LGTM

**Pull request merged**